mypy/moduleinspect.py

```diff
@@ -0,0 +1,160 @@
+                            is_c_module=is_c,
+                            subpackages=subpackages)
+
+
+def worker(queue1: 'Queue[str]', queue2: 'Queue[Union[str, ModuleProperties]]') -> None:
+    """The main loop of a worker introspection process."""
+    while True:
+        mod = queue1.get()
+        try:
+            prop = get_package_properties(mod)
+        except InspectError as e:
+            queue2.put(str(e))
+            continue
+        queue2.put(prop)
+
+
+class ModuleInspect:
+    """Perform runtime introspection of modules in a separate process.
+
+    Reuse the process for multiple modules for efficiency. However, if there is an
+    error, retry using a fresh process to avoid cross-contamination of state between
+    modules.
+
+    We use a separate process to isolate us from many side effects. For example, the
+    import of a module may kill the current process, and we want to recover from that.
+
+    Always use in a with statement for proper clean-up:
+
+      with ModuleInspect() as m:
+          p = m.get_package_properties('urllib.parse')
+    """
+
+    def __init__(self) -> None:
+        self._start()
+
+    def _start(self) -> None:
+        self.q1 = Queue()  # type: Queue[str]
+        self.q2 = Queue()  # type: Queue[Union[ModuleProperties, str]]
+        self.proc = Process(target=worker, args=(self.q1, self.q2))
+        self.proc.start()
+        self.counter = 0  # Number of successfull roundtrips
+
+    def close(self) -> None:
+        """Free any resources used."""
+        self.proc.terminate()
+
+    def get_package_properties(self, package_id: str) -> ModuleProperties:
+        """Return some properties of a module/package using runtime introspection.
+
+        Raise InspectError if the target couldn't be imported.
+        """
+        self.q1.put(package_id)
+        res = self._get_from_queue()
+        if res is None:
+            # The process died; recover and report error.
+            self._start()
+            raise InspectError('Process died when importing %r' % package_id)
+        if isinstance(res, str):
+            # Error importing module
+            if self.counter > 0:
+                # Also try with a fresh process. Maybe one of the previous imports has
+                # corrupted some global state.
+                self.close()
+                self._start()
+                return self.get_package_properties(package_id)
+            raise InspectError(res)
+        self.counter += 1
+        return res
+
+    def _get_from_queue(self) -> Union[ModuleProperties, str, None]:
+        """Get value from the queue.
+
+        Return the value read from the queue, or None if the process unexpectedly died.
+        """
+        max_iter = 100
+        n = 0
+        while True:
+            if n == max_iter:
+                raise RuntimeError('Timeout waiting for subprocess')
+            try:
+                return self.q2.get(timeout=0.05)
```

Comment:
Maybe make the timeout configurable in constructor?

Reply:
I don't think that making it configurable is useful.